### ISSUE TITLE
Visual Studio runner adaptor fails if the target test assembly doesn't use xUnit

### ISSUE THREAD

**Issue by jdom** [xunit/xunit#746]
Hi, I have a legacy solution where we are very slowly migrating to xUnit from MSTest, so some of the test assemblies use xUnit and some MSTest. Most of our builds compiles the assemblies and puts them all together in the same directory.
If only some of the assemblies target xUnit, then those assemblies run their tests correctly. Nevertheless when we try to run the tests in our other (non-xUnit) test assemblies, because the xunit.dll and the adaptor are present in that same folder, then  `vstest.console` tries to discover tests with xUnit too.
xUnit at this point logs an Error (that I have no way of turning into a Warning, so it fails my builds):

```
[xUnit.net 00:00:03.6449324] MyCustomTestsWithMsTest: Exception discovering tests: No tests matched the filter because it contains one or more properties that are not valid (TestCategory). Specify filter expression containing valid properties (DisplayName, FullyQualifiedName) and try again. 
```

The error comes from the following line:
[TestCaseFilterHelper.cs#L94](https://github.com/xunit/visualstudio.xunit/blob/master/xunit.runner.visualstudio.testadapter/Utility/TestCaseFilterHelper.cs#L94)

Ideally, this assembly should have been skipped because it doesn't reference xUnit at all, but looking at [VsTestRunner.cs#L300](https://github.com/xunit/visualstudio.xunit/blob/master/xunit.runner.visualstudio.testadapter/VsTestRunner.cs#L300) it seems like that is not something that is checked, and all that matters is that xUnit is in the same folder

**Comment by scripni** [comment]
See if you can fix this in the meantime by switching to v2.0.1, according to issue  #610 that version doesn't have this issue.

Would be nice to have a fix moving forward though.

**Comment by jdom** [comment]
Yeah, I saw that and tried it, but with that version I couldn't get filtering through traits to work. I also tried the latest 2.2.0-beta and 2.1.0 worth no luck.
In the meantime, as a dirty workaround, I'm segregating builds so that the ones that run mstest, they first delete every xunit binary from the bin folder. Thanks for the suggestion though

**Comment by cobysy** [comment]
I found a workaround though: just add a placeholder class decorated with the trait that you're filtering with (and add a Fact):

```csharp
using Xunit;

namespace xunitissue2
{
    public class Class1
    {
        [Fact]
        public void X()
        {
            Assert.True(false);
        }
    }


    /// <summary>
    /// Workaround so that xUnit2 can detect tests in this assembly when a filter expression is specified.
    /// </summary>
    [Trait("Category", "Integration")]
    public class _
    {
        [Fact]
        public void __()
        {
        }
    }
}
```

But experienced the same and tried switching back to **1.9.2** but `vstest.console` **hangs** instead leaving a hanging `te.processhost.managed.exe`.

```
<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="xunit" version="1.9.2" targetFramework="net452" />
  <package id="xunit.runner.visualstudio" version="2.1.0" targetFramework="net452" />
</packages>
```

```csharp
using Xunit;

namespace xunitissue2
{
    public class Class1
    {
        [Fact]
        public void X()
        {
        }
    }
}
```

**Comment by brunobertechini** [comment]
+1 Same problem here

**Comment by bradwilson** [comment]
Does this still repro with 2.2 beta 3?

**Comment by brunobertechini** [comment]
Im not sure. Im using 2.1.0 as we can not upgrade it now...(no access to build servers at this moment)

**Comment by jsblock78** [comment]
I see this issue in 2.2 beta 3. The workaround provided by @cobysy works for me, but is messy.

**Comment by danstiner** [comment]
In short the issue is a test case filtering API being used requires a list of valid trait names, but it is called with the list of trait names seen _so far_ as discovery sequentially steps through assemblies. The fix I think inevitably involves gathering the list of trait names in all assemblies before starting to filter anything.

In a bit more detail, the issue is with the VSTest TestCaseFilter API that requires a list of supported properties be passed in when filtering. During test case discovery (calls to DiscoverTests), each assembly is reflected over for test cases which are then passed to the filter API, using the so far accumulated set of observed trait names as the supported properties. So if any property in the filter is not defined by a trait in the first assembly processed, Visual Studio's TestCaseFilter API will throw meaning no tests will discovered or run from that assembly. The exception does get caught, but the overall execution result is still a fail.

Relevant VSTest filtering API:
```
ITestCaseFilterExpression GetTestCaseFilter(IEnumerable<String> supportedProperties, Func<string, TestProperty> propertyProvider);
```
Detailed example at: https://blogs.msdn.microsoft.com/vikramagrawal/2012/07/23/running-selective-unit-tests-in-vs-2012-rc-using-testcasefilter/

The issue is related to how trait names are accumulated in `knownTraitNames`: https://github.com/xunit/visualstudio.xunit/blob/master/xunit.runner.visualstudio.desktop/VsTestRunner.cs#L274

I've prototyped an approach of modifying discovery to find all test cases before doing any filtering. The proposed workflow would be:
 1. Discover all test cases in all assemblies into a single data structure
 2. Extract set of all trait names
 3. Do filtering
 4. Unique-ify test case names as currently is being done to satisfy #603 

In contrast the current workflow is:
 1. Start with empty set of known trait names
 2. For each assembly:
   1. Discover test cases
   2. Extract set of trait names and add to known set
   3. Do filtering using current known set
   4. Unique-ify test case names per #603 

Happy to send a pull request. It was a couple months ago but I did have a working prototype.
